sway-lib-std/src/vec.sw

```diff
@@ -0,0 +1,189 @@
+library vec;
+
+use ::context::registers::heap_ptr;
+use ::option::Option;
+
+struct RawVec<T> {
+    ptr: u64,
+    cap: u64,
+}
+
+impl<T> RawVec<T> {
+    /// Create a new `RawVec` with zero capacity.
+    fn new() -> Self {
+        RawVec {
+            // Heap pointer points to _unallocated_ memory.
+            ptr: heap_ptr() + 1,
+            cap: 0,
+        }
+    }
+
+    /// Creates a `RawVec` (on the heap) with exactly the capacity for a
+    /// `[T; capacity]`. This is equivalent to calling `RawVec::new` when
+    /// `capacity` is `0`.
+    fn with_capacity(capacity: u64) -> Self {
+        asm(size: capacity * __size_of::<T>()) {
+            aloc size;
+        };
+        RawVec {
+            // Heap pointer points to _unallocated_ memory.
+            ptr: heap_ptr() + 1,
+            cap: capacity,
+        }
+    }
+
+    /// Gets the pointer of the allocation.
+    fn ptr(self) -> u64 {
+        self.ptr
+    }
+
+    /// Gets the capacity of the allocation.
+    fn capacity(self) -> u64 {
+        self.cap
+    }
+
+    fn grow(self) {
+        let mut self = self;
+        let new_cap = if self.cap == 0 {
```

Comment:
no match statement?

Reply:
This code is copied from John's implementation here: https://github.com/FuelLabs/sway/blob/906903aa247d21f1fdf6334e49f16f54c6b40372/sway-lib-std/src/vec.sw